src/read_write/node_iterator.rs

```diff
@@ -2,178 +2,192 @@ use crate::codec::{decode, fixpoint_decode};
 use crate::color;
 use crate::errors::*;
 use crate::math::Cube;
-use crate::octree::{
-    NodeId, NodeLayer, NodeMeta, OctreeDataProvider, OctreeMeta, PositionEncoding,
-};
+use crate::octree::{NodeLayer, PositionEncoding};
 use crate::Point;
 use byteorder::{LittleEndian, ReadBytesExt};
 use cgmath::Vector3;
 use num_traits::identities::Zero;
+use std::collections::HashMap;
 use std::io::{BufReader, Read};
 
-/// Streams points from our data provider representation.
-pub struct NodeIteratorWithData {
-    xyz_reader: BufReader<Box<dyn Read>>,
-    rgb_reader: BufReader<Box<dyn Read>>,
-    intensity_reader: Option<BufReader<Box<dyn Read>>>,
-    meta: NodeMeta,
-    point_count: usize,
-}
-
-pub enum NodeIterator {
-    WithData(NodeIteratorWithData),
-    Empty,
+pub trait NodeReader {
+    fn read(&mut self) -> Point;
 }
 
-impl NodeIterator {
-    pub fn from_data_provider(
-        octree_data_provider: &dyn OctreeDataProvider,
-        octree_meta: &OctreeMeta,
-        id: &NodeId,
-        num_points: i64,
-    ) -> Result<Self> {
-        if num_points == 0 {
-            return Ok(NodeIterator::Empty);
-        }
-        let bounding_cube = id.find_bounding_cube(&Cube::bounding(&octree_meta.bounding_box));
-        let position_encoding = PositionEncoding::new(&bounding_cube, octree_meta.resolution);
-        let intensity_reader = match octree_data_provider.data(id, vec![NodeLayer::Intensity]) {
-            Ok(mut data_map) => match data_map.remove(&NodeLayer::Intensity) {
-                Some(intensity_data) => Some(BufReader::new(intensity_data)),
-                None => {
-                    return Err("No intensity reader available.".into());
-                }
-            },
-            Err(_) => None,
-        };
-
-        let mut position_color_reads =
-            octree_data_provider.data(id, vec![NodeLayer::Position, NodeLayer::Color])?;
-        let iter = NodeIteratorWithData {
-            xyz_reader: BufReader::new(
-                position_color_reads
-                    .remove(&NodeLayer::Position)
-                    .ok_or_else(|| "No position reader available.")?,
-            ),
-            rgb_reader: BufReader::new(
-                position_color_reads
-                    .remove(&NodeLayer::Color)
-                    .ok_or_else(|| "No color reader available.")?,
-            ),
-            intensity_reader,
-            meta: NodeMeta {
-                bounding_cube,
-                position_encoding,
-                num_points,
-            },
-            point_count: 0,
-        };
-        Ok(NodeIterator::WithData(iter))
-    }
+pub struct CubeNodeReader {
+    readers: Vec<BufReader<Box<dyn Read>>>,
```

Comment:
Maybe the xyz reader should be a separate member. That would enforce that the position is always available and make the code a little more readable.